Layers window updates

dsizzle opened the pull request:
I moved the layer operations up to the top and out of each layer, and added a transparency slider so you can adjust layer transparency easily. Also re-enabled(?) the layer name field and moved visibility to the left.  This is more like how Photoshop's layer panel is. 

oh and the changes to the rendering are in preparation for adding the ability to change to different compositing modes for the layers. 

However, the layouting is giving me headaches.  I'm positing this mostly for comments or suggestions - it's not done but wanted to get thoughts. 

Related to #109

dsizzle commented:
ok, I think I got the layouting fixed now.  

But is there something wrong with the CI build?  The message says it can't find 'make' ??

dsizzle commented:
oh, also Fixes #20

humdingerb commented:
Sorry by the late feedback.
A few suggestions:
- Add a horizontal separator under the file name.
- The transparency setting takes quite a lot of space. I think we may get away with a label "α" (= alpha symbol), the slider and a numbers box (just wide enough to fit "100"), all in one row.
- The "Action" menu could be moved into a regular menubar at the top. I expect we'll use the space for a "Blending mode" menu at one time.
- The down-arrow of the scrollbar is behind the resize-corner.
- It'd be nice if the window were resizable horizontally and made a bit wider by default.
- The borders of the box above the layers should merge with the window borders.
- Very long filenames stretch the window, instead the filename should be middle-truncated.

Kudos for tackling the layers window, though! :)

dsizzle commented:
I think all of your comments are addressed, except the horizontal scaling - it really messes things up to allow that.  I did make the window a little wider. 

@korli do you know why the CI builds are failing?  it complains about not being able to find the 'make' command?

humdingerb commented:
I thought I tested your last push, did a clean build... Anyway, now the scroll arrow is indeed visible.

More observations:
- The whole scrollbar needs to move 1 pixel to the right to avoid a "thick" border with the window border.
- The scrollbar behaves a bit strange, e.g. it gets a scroll-knob when resizing the window, even if it still isn't needed. Even if you have just one layer, after making the window bigger and smaller again, you can scroll the layer out of the visible area.
- For some reason, transparency isn't working. It's all black...
- The file name is replaced with random symbols when adding a layer.
- The layers are sorted the wrong way round. "Add layer in front" ends with a layer below the current layer. (I always wondered, if we should go from "front/behind" to "above/below". It fits better to the Layers window and is equally real-worldy, considering a stack of sheets.)
- I also wonder if we need to add/merge layers "in front" and "below". We could just "Add layer" and "Merge layer" and always just "add above" and "merge with below". Saves menu items, and may come in handy if we have icon-buttons for all actions for quicker access.

dsizzle commented:
> For some reason, transparency isn't working. It's all black...

 Which part is black? The canvas is black? It works perfectly on my end.

humdingerb commented:
Yep. Anything translarency related. If I launch ArtPaint and create a new canvas, it's black.
The only place I get the alpha-chckers is in the Layers window. I have completely removed all objects folders and built from scratch.

humdingerb commented on artpaint/layers/LayerView.cpp:
```diff
@@ -28,76 +31,30 @@
+		.Add(visibility_box)
+		.Add(thumbnail_view)
+		.Add(layer_name_field)
+		.SetInsets(5.0, 5.0, 5.0, 5.0);
```
It may be a good idea to use constants from the InterfaceDefs.h for all insets, e.g. B_USE_SMALL_INSETS.

humdingerb commented on artpaint/layers/LayerView.cpp:
```diff
@@ -112,33 +69,39 @@ LayerView::~LayerView()
-		a_rect.OffsetTo(4,4);
-		SetHighColor(0,0,255,255);
+		BRect a_rect = thumbnail_view->Frame();
+		SetHighColor(0, 0, 255, 255);
```
Better use ui_color(B_CONTROL_HIGHLIGHT_COLOR). If that's the right constant...

humdingerb reviewed (commented):
Maybe my transparency issues stem from changes in paintwindow/Image.cpp ?

dsizzle commented:
I did change the main rendering loop, which is where we started this whole game - I think my very first fix in ArtPaint was the endianness fix for the black canvas. Maybe I will back out that part of the change for now. I am quite grateful for your testing!!

dsizzle commented:
Fixed some but not all:

- fixed filename truncation so it doesn't give weird symbols
- fixed layer order
- moved scroll bar over by 1 pixel

I still haven't seen the black canvas issue which is concerning. I'll still pull that change out but wish I could reproduce it because I want that change for the layer compositing modes.

humdingerb commented:
HA!
I found out how to the transparency working again, but I have no idea why that is.
I had all "Keep in front" checkmarks in the "Windows" settings disabled. After ticking a checkmark and pressing OK, the transparency was back; checkers and all.
Even after removing the checkmark again, the checkers stayed.
It's back to black, if I delete all settings.

humdingerb commented:
Oh yeah, your other fixes are all good. The scrollbar, though, needs to grow 1 pixel at the top and bottom, I think. The Magnify app is an unforgiving tool... :)

humdingerb commented:
The Layers window isn't resizable anymore.

dsizzle commented:
THANKS for digging into that!!  If I delete the settings I am able to reproduce the black canvas problem.  I submitted the fix as a separate PR as it doesn't have anything to do with this change and I think it should be pulled into the mainline separately. (#227)

Scroll bar / resizing still in progress...

dsizzle commented:
Ok, I fixed the resizing, and moved the scrollbar down 1 pixel.  I can't seem to change the top pixel for some reason, perhaps because of how the top and bottom layout butt against each other?

Also still need to figure out the scroll bar still.

humdingerb commented:
I still cannot manually resize the Layers window.

Which also rises a question: Can we somehow manage to simultaneously resize the window automatically, and respect a manual resize?
On the one hand, it's nice to make the Layers window as small as possible and auto-grow the height when a new layer is added. (Shouldn't grow outside the screen, of course.)
On the other hand, users may want to limit the window height and rather scroll up/down, in order not to obscure too much of the canvas.

Maybe ArtPaint can remember the window height it was last _manually_ resized to. Keep the window height as small as possible while auto-growing/shrinking with additional/removed layers. When the previous manual-height is reached, stop auto-growing.

Sounds reasonable?
Something for later, i.e. create an enhancement ticket?

korli commented:
The CI image is rebuilt [every week or so 1](https://github.com/hectorm/docker-qemu-haiku/issues/1#issuecomment-1086655709). Please be patient, or ignore this for a few days :)

dsizzle commented:
I just deleted everything (including the settings) and rebuilt, and resizing still works for me. There is a minimum size for the window, but otherwise it should resize  - see this horrific example:

![1layer](https://user-images.githubusercontent.com/10878750/182288507-91899fd4-8b12-42f4-ba33-9bb6f4ba3b21.png)

You can file the enhancement for resizing as you describe but it sounds like a challenging bit of work.

dsizzle commented:
Ok, after a lot of head banging and a desperate reachout in the forums and then more head banging I got the scrollbar to work properly.  Whew.  

Mostly.  It doesn't update the scrollbar when you add a new layer.  But otherwise it works as it should.  If I can figure out the update-on-new-layer then I *think* that's the last piece here.

oh, i just remembered the "above/below" change too.

dsizzle commented:
ok, I figured out how to make the scrollbar update.

Also changed the language to use "above" and "below" and I think it makes way more sense now.  

I think this is all working; please give a try when you have a moment and I'm sure you'll find things to break. :laughing: 

also waiting on the CI build to work again...

dsizzle commented:
I am working on the layer name saving/loading; I feel like that should be part of this since I'm showing the names now.  

Also I think drag-to-reorder is a bit broken...will fix that too. 

Once I fix those I'm ok not waiting for the CI bot if you are ok with that.  Just didn't want to break protocol.

korli commented:
@dsizzle please rebase this PR on master to get the CI update. Thanks.

dsizzle commented:
rebased to get CI fix, thanks @korli !

hmm. I fixed the visual of layer reordering, but if you resize the window the layers revert to the original order.  I must've broken something with the layer ordering somewhere along the way. :frowning_face:

humdingerb commented:
It's really quite weird. The layer mayhem is not just visible in the Layers window, but is seem in the canvas as well. (You're most probably very aware of this, but I thought I mention it, because it read like you were focusing solely on the Layers window.)

Good luck hunting that bug!

dsizzle commented:
Thanks, yes, I was aware of the canvas having the same mayhem - good word for it, btw!

I think I fixed the overall window behavior for dragging, but somehow stuff still gets out of order in the canvas?  but - at least taking steps toward the goal.

dsizzle commented:
Ok, I think I solved it!! Please give it a try when you have a chance, I believe all layer reordering should work as expected now. 

I also re-added disabling the merge menu items when there aren't layers above or below, and added a tooltip to the slider.  Unfortunately due to the layouting I couldn't figure out how to get the tooltip on anything except the slider itself.

humdingerb commented on artpaint/layers/LayerWindow.cpp:
```diff
@@ -46,28 +63,87 @@ sem_id LayerWindow::layer_window_semaphore = create_sem(1,"layer window semaphor
+	message->AddInt32("value", 0);
+
+	transparency_slider =
+		new NumberSliderControl(B_TRANSLATE("\xCE\xB1:"), "0",
```
The alpha symbol shouldn't need to be translatable. Without a comment, nobody will guess what it is anyway. :)

humdingerb commented on artpaint/layers/LayerWindow.cpp:
```diff
@@ -46,28 +63,87 @@ sem_id LayerWindow::layer_window_semaphore = create_sem(1,"layer window semaphor
+	transparencyLayout->SetMinColumnWidth(2,
+		font.StringWidth("SLIDERSLIDERSLIDER"));
+
+	transparency_slider->Slider()->SetToolTip(B_TRANSLATE("Set Layer Transparency"));
```
Sentence casing: "Set layer transparency" or maybe just "Layer transparency" as the whole widget also just shows the layer transparency, if the user doesn't twiddle the slider knob.

humdingerb commented on artpaint/layers/LayerWindow.cpp:
```diff
@@ -143,9 +260,9 @@ void LayerWindow::ActiveWindowChanged(BWindow *active_window,BList *list,BBitmap
 	}
 
-	if (a_layer != NULL)
+	if (a_layer != NULL) {
```
No need for the curly brackets.

humdingerb commented on artpaint/layers/LayerWindow.cpp:
```diff
@@ -198,19 +316,24 @@ LayerWindow::setFeel(window_feel feel)
 		if (feel == B_NORMAL_WINDOW_FEEL) {
-			layer_window->SetLook(B_DOCUMENT_WINDOW_LOOK);
-			layer_window->scroll_bar->ResizeTo(layer_window->scroll_bar->Bounds().Width(),layer_window->list_view->Bounds().Height()-B_H_SCROLL_BAR_HEIGHT+1);
+			layer_window->SetLook(B_TITLED_WINDOW_LOOK);
```
No need for the curly brackets.

humdingerb commented on artpaint/layers/LayerWindow.cpp:
```diff
@@ -238,68 +362,84 @@ void LayerWindow::Update()
-			layer_window->list_view->RemoveChild(layer_window->list_view->ChildAt(0));
+		BGridLayout* layout = (BGridLayout*)layer_window->list_view->GetLayout();
+
+		while (layout->CountItems() > 0) {
```
No need for the curly brackets.

humdingerb commented on artpaint/paintwindow/ImageView.cpp:
```diff
@@ -367,6 +367,11 @@ void ImageView::MessageReceived(BMessage *message)
 			}
 		}	break;
 
+		case HS_LAYER_TRANSPARENCY_CHANGED:
```
The other cases do have curly brackets...

dsizzle replied:
I left this one alone for now; about half of the case HAVE curly braces and half don't.  That file needs a cleanup overall.

humdingerb commented on artpaint/layers/LayerView.cpp:
```diff
@@ -351,9 +306,57 @@ int32 LayerView::ReorderViews()
+
+ThumbnailView::~ThumbnailView()
+{
+
```
Remove empty line.

humdingerb reviewed (commented):
Just a few minor comments.

humdingerb commented:
The layer ordering works very well now. Congrats for seeing this through! :)

What do you think about keyboard shortcuts for the Action menu items? Like:
* ALT+CursorLeft for  "Merge with layer above"
* ALT+CursorRight for  "Merge with layer below"
* ALT+CursorUp for  "Add layer above"
* ALT+CursorDown for  "Add layer below"
* ALT+D for "Duplicate layer"
* ALT+T for "Delete layer"

dsizzle commented:
Thanks!  

I added the keyboard shortcuts, except I changed these two:
- ALT+SHIFT+CursorUp for "Merge with layer above"
- ALT+SHIFT+CursorDown for "Merge with layer below"

My reasoning was twofold: one, to keep with up and down arrows for above and below, and also adding shift makes the merging more deliberate and less likely to confuse with adding a new layer.

dsizzle commented:
Fixes #109 btw

humdingerb commented:
Makes sense.
So, how do you want to play it: Just merge #227 and build a quick bugfix hpkg? Or merge both #227 and this one, put the changes on Polyglot, wait a bit, then make a proper release?

humdingerb commented:
I found one little detail, just if you're not already fed up with the layer stuff...

Say you have 3 layers: L1, L2, L3. You select L2 and merge with above (L1), everything is perfect.
But if you merge with below (L3), L1 becomes the selected layer, instead of staying the now combined L2+L3.

dsizzle commented:
The benefit of merging just #227 with a quick fix is that we could also do a polyglot export. I think there are a few new languages. I’m not sure this layer change really justifies a release by itself, and it requires polyglot work for like 1 new string. Might wait until more features are added for a “full” release. Does this make sense? 

I would like to merge this change after that, in case I tire of layers and want to work on other stuff. I’m not sure if layer merging in general is correct/intuitive; if you merge L2 into L1 would you expect the layer to be called L1 or L2? It seemed to be the opposite of what I expected when I was playing with it.

humdingerb commented:
> Might wait until more features are added for a “full” release. Does this make sense?

Alright. Bugfix + translations it is.

> I’m not sure if layer merging in general is correct/intuitive; if you merge L2 into L1 would you expect the layer to be called L1 or L2? It seemed to be the opposite of what I expected when I was playing with it.

[Just for clarification, above I talk about what layer is selected after a merge, not the layer label.]
One could complicate things and distinguish if you're merging up or down, but IMO it's best to just be consistent and just keep the label of the layer that was selected when the merge was initiated.

humdingerb commented:
Arggl... I messed up on Polyglot by accidentally uploading the en.catkeys of this PR. That added the new strings and removed the old ones in all languages...  :(

Oh well, then we have just a quick bufix release and no new translations. Those come with the new layers window and whatever else you sprinkle your magic coder dust on.

dsizzle commented:
Yep, I see what you mean with the selecting the wrong layer when using "merge with below"

btw, ALT+T is already "Translate Layer" and ALT+CursorLeft/Up are shortcuts for flipping layers horizontally and vertically.  It was rather concerning when i was trying to move layers and they kept disappearing! :laughing:  Any other ideas for keyboard shortcuts?

also noticed that the Undo/Redo menu text still says "Merge with back/front layer" so I will fix that.

and finally "magic coder dust" is what's left when coders run out of tears. :smile:

humdingerb commented:
> btw, ALT+T is already "Translate Layer" and ALT+CursorLeft/Up are shortcuts for flipping layers horizontally and vertically. It was rather concerning when i was trying to move layers and they kept disappearing! 😆 Any other ideas for keyboard shortcuts?

Damn the scarcity of good shortcuts!
Truth be told, considering the small number of "actions" in that menu and that, as a graphics app, most users are quite mouse-centered, it may be best to use a tool bar with icon buttons instead.
We can whittle down the number of actions further by only doing a "merge down" and a "add layer above". Then we only need 4 buttons: add, delete, merge, duplicate.

I also experienced my shortcuts being eaten by focused "sub-windows". It'd be nice to have all shortcuts being sent to the main window, if they don't apply to a sub-window menu item (which should generally be avoided IMO, exception would be a to be added "DEL" press in the "Brushes" window to delete the current brush).
Is that possible?

> and finally "magic coder dust" is what's left when coders run out of tears. 

Oh dear. :)

BTW, whenever you want to draw a line under this PR, give the word.

dsizzle commented:
> Truth be told, considering the small number of "actions" in that menu and that, as a graphics app, most users are quite mouse-centered, it may be best to use a tool bar with icon buttons instead.
We can whittle down the number of actions further by only doing a "merge down" and a "add layer above". Then we only need 4 buttons: add, delete, merge, duplicate.

yeah, I initially made buttons but figured I'd stick with a menu for now.  I'm not sure how I feel about decreasing the number of actions as it's probably nicer to be able to add/merge in both directions.  

I also find myself trying to right-click to add/duplicate/merge so i wonder if having the menu pop up on right click would make sense even with buttons.

that said, there really SHOULD be a "merge selected" and "merge all" (aka "flatten image" in Photoshop-speak).  Currently this window only supports a single selection and really someday it should handle multi-select of layers for various kinds of actions

>  It'd be nice to have all shortcuts being sent to the main window...Is that possible?

Idk. it should be. There are currently unfortunate side effects to the way it works now.  For example if you merge layers and want to undo, you have to click on the main canvas. "undo" with the layer window selected does nothing. 

> BTW, whenever you want to draw a line under this PR, give the word.

I removed the shortcuts for now and I fixed the layer selection when you merge, so unless you find something broken I would be happy to see this merged. Thanks!

humdingerb commented:
> I also find myself trying to right-click to add/duplicate/merge so i wonder if having the menu pop up on right click would make sense even with buttons.

I'd prefer a context menu, too. Even more so currently, as all 'actions' relate to the selected layer. It's natural to right-click the layer to manilpulate and have all 'actions' right there.
I'm not sure sure I'd use it were there icon-buttons, but in case we get more 'actions' in the future, the menu would be a keeper.

> I'm not sure how I feel about decreasing the number of actions as it's probably nicer to be able to add/merge in both directions. [...]
> that said, there really SHOULD be a "merge selected" and "merge all" (aka "flatten image" in Photoshop-speak). Currently this window only supports a single selection and really someday it should handle multi-select of layers for various kinds of actions

Agreed to the multi-selection features. In case of merging multiple layers, the merging would be in one direction, i.e. merging with the below (the upper layer staying on top the ones below). I think it's OK to train the user that there's just one direction to merge: down.
I'm trained that way by Photoshop with it's CTRL+E shortcut when merging...

Anyway, I'll fire up the 32bit Haiku to check the build there and then merge.
The Layers window is much nicer now, and more functional. Well done!!

humdingerb commented:
Will merge after the 32bit buildfix. :P

dsizzle commented:
What's broken with the 32-bit build?  Do I need to fix something?

dsizzle commented:
sorry about that - should be fixed now.

humdingerb commented:
Thanks. Let's merge!
